# Project Title
Comparison of Logistic Regression and SVM Method

## Members

1. First member: name and e-mail adress
Zeynep Göger and gogerz@itu.edu.tr
2. Second member: name and e-mail adress

## Description of the project
In this project,  we compare two methods that are logistic regression method and SVM method 
by using 1984 United Stated Congressional Voting Records in R language. We analyse that which
method is better to predict the congressman belongs to which party according to first column
of the data to the rest of them.


### The methods to be used
There are two different methods that are logistic regression method and SVM method.
Logistic regression is a type of a prediction method that is used if the data has binary 
dependent variables and this method determines the relationship between the set of dependent
variables and the independent set.
    
Support Vector Machines method is another prediction method that is used if the data is
binary too. However, in this method two kinds of accumulation of dependent variables
are determined and divided down the middle of the distance that is called as margin.    

Explain.

### The data

The data is 1984 United Stated Congressional Voting Records that has 435 rows and 17 columns. This data consists of 435 congressmen that are 267 democrats and 168 republicans.
In this data has binary dependent variables such as yes or no and two class of congressmen
that are republican and democrat. There are 16 numbers of attributes that are categorical.
    
## Code

As proof of work, you must run this notebook.  Upload an HTML output of this notebook on your github account.

In [1]:
X<- read.csv("https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data", header=FALSE)

In [2]:
head(X)

V1         V2 V3 V4 V5 V6 V7 V8 V9 V10 V11 V12 V13 V14 V15 V16 V17
1 republican n  y  n  y  y  y  n  n  n   y   ?   y   y   y   n   y  
2 republican n  y  n  y  y  y  n  n  n   n   n   y   y   y   n   ?  
3 democrat   ?  y  y  ?  y  y  n  n  n   n   y   n   y   y   n   n  
4 democrat   n  y  y  n  ?  y  n  n  n   n   y   n   y   n   n   y  
5 democrat   y  y  y  n  y  y  n  n  n   n   y   ?   y   y   y   y  
6 democrat   n  y  y  n  y  y  n  n  n   n   n   n   y   y   y   y

## Logistic Regression
Logistic regression is a type of a prediction method that is used if the data has binary 
dependent variables and this method determines the relationship between the set of dependent
variables and the independent set. Odds is the ratio of the probability of occurrence with probability of event that does not occur.

$logit_p = log(odds) = log(p/(1-p)) =  b_0 + b_1x_1 +...+  b_k x_k$

where $b_k's$ are constant coefficients that determines the changes of logistic regression model when $x_i$ is added.

Initially, we divide the data to constitute our model and we divide 30% of the data randomly.


To do this we use this code 


In [3]:
sample(1:nrow(X), 0.3*nrow(X))
test <-sample(1:nrow(X), 0.3*nrow(X))

[1] 234 242  60  82 164  40 291 149 155 203 337  23 312  68 201 273  96 324
 [19] 165  57 290 174 151 249 272 364 196 352 228 295 233  26  71 134  20 263
 [37] 121 359 265 256 167 202  95 124  99  98 247  77 354  31 303   8 245 282
 [55]  62  43 336 329 425 148 237  32 230 145 232 138 133 171  16 117 368  84
 [73]   2 396 104 262 229 393  10 270  94 308 392 391 380 301 309 418 143 345
 [91] 110  42 406 198 340 131 408 433 120 401 114 422 205  92 333 172 306 395
[109] 296 144   4 101 207 399 208  37  66  97 146 222  47 322 236 367 307 107
[127] 432 254  50  90

In [4]:
length(test)


[1] 130

130 members of this data is selected randomly by using this code in R but these numbers are not useful for testing and forming our model. Therefore, we convert them to a matrix by using the code below.

In [5]:
X[test,]

V1         V2  V3  V4  V5  V6  V7  V8  V9  V10 V11 V12 V13 V14 V15 V16 V17
160 democrat   n   y   y   n   ?   y   y   y   y   y   y   n   n   ?   n   ?  
213 democrat   n   n   y   n   n   n   y   y   y   n   y   n   n   n   y   y  
130 democrat   ?   ?   y   n   n   n   y   y   ?   n   ?   ?   ?   ?   ?   ?  
279 republican n   n   n   y   y   y   n   n   n   n   n   y   y   y   n   y  
162 democrat   n   n   n   n   y   y   y   n   n   n   n   y   y   y   n   y  
276 democrat   n   n   y   n   y   y   n   n   n   n   ?   n   y   y   n   n  
182 democrat   n   n   y   n   n   n   y   y   y   y   y   n   n   n   y   y  
78  democrat   n   y   y   y   y   y   n   y   y   y   y   y   y   y   n   y  
335 democrat   n   y   y   n   n   y   y   y   ?   n   y   y   n   n   y   y  
382 democrat   y   y   y   n   y   y   y   y   y   y   y   n   y   y   n   ?  
418 democrat   y   y   y   n   n   n   y   y   y   y   y   n   y   n   n   y  
254 republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   y  
58  republican n   y   n   y   y   y   n   n   n   y   n   y   y   y   n   y  
251 republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   n  
165 democrat   y   y   n   n   y   y   n   n   n   y   y   y   y   y   n   ?  
27  democrat   y   n   y   n   n   n   y   y   y   n   y   n   n   n   y   y  
222 democrat   n   y   y   n   y   y   n   n   n   n   y   n   n   n   y   ?  
181 democrat   ?   ?   y   n   n   n   y   y   y   ?   ?   n   n   n   ?   ?  
411 republican n   n   n   y   y   y   n   n   n   n   y   y   y   y   n   y  
290 democrat   y   y   y   n   y   y   n   y   n   y   y   n   y   y   n   ?  
62  republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   n  
297 republican n   n   y   y   y   y   n   n   n   y   n   y   y   y   y   y  
220 democrat   n   y   y   n   n   n   n   y   y   n   y   n   n   y   y   y  
113 democrat   n   ?   y   n   n   y   y   y   n   y   n   n   n   n   y   ?  
302 democrat   n   n   ?   n   n   y   y   y   y   n   n   n   n   n   y   y  
203 democrat   y   n   y   n   n   y   y   y   y   y   y   n   n   n   y   y  
275 republican y   n   n   y   y   n   y   n   n   y   n   n   n   y   y   y  
132 democrat   n   n   y   n   n   y   n   y   y   y   n   n   n   y   n   y  
322 democrat   y   y   y   n   n   n   y   y   y   y   n   n   y   n   n   y  
239 democrat   y   y   y   n   n   n   y   y   ?   y   n   n   n   n   y   ?  
... ...        ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ...
221 democrat   n   n   y   n   n   n   y   y   y   n   n   n   n   n   y   ?  
386 democrat   y   y   n   n   y   y   n   n   n   n   y   y   y   y   y   n  
119 democrat   n   y   y   n   n   n   n   y   y   y   y   n   n   n   y   y  
435 republican n   y   n   y   y   y   n   n   n   y   n   y   y   y   ?   n  
303 republican n   n   n   y   y   y   y   n   n   y   n   y   y   y   n   y  
111 democrat   n   n   y   n   n   n   y   y   y   y   n   n   n   n   y   y  
258 republican n   n   n   y   y   n   n   n   n   n   n   y   n   y   ?   y  
85  republican n   ?   n   y   y   y   n   n   n   y   n   y   y   y   n   n  
120 republican n   n   n   y   y   y   n   n   n   n   n   y   y   y   n   n  
186 democrat   y   n   y   n   n   n   y   y   y   y   n   ?   n   n   y   y  
159 republican n   y   n   y   y   y   n   n   n   y   n   y   y   y   n   n  
336 republican n   n   n   y   y   y   n   n   n   y   y   y   y   y   n   ?  
272 democrat   n   y   y   n   n   y   y   y   y   n   ?   n   n   n   n   y  
333 democrat   y   n   y   n   n   n   y   y   y   n   y   n   n   n   y   ?  
97  democrat   n   n   ?   n   y   y   n   n   n   n   y   y   y   y   n   y  
349 democrat   y   n   y   n   n   n   y   y   y   y   n   y   n   n   y   ?  
422 democrat   n   y   y   n   y   y   y   y   n   n   y   n   y   n   y   y  
99  democrat   y   y   y   n   n   y   y   y   y   y   n   n   n   n   n   y  
354 republican n   n   y   y   y   y   y   n   n   y  

In [6]:
nrow(X[test,])
modeldata<-X[test,]

[1] 130

In [7]:
X[-test,]

V1         V2  V3  V4  V5  V6  V7  V8  V9  V10 V11 V12 V13 V14 V15 V16 V17
1   republican n   y   n   y   y   y   n   n   n   y   ?   y   y   y   n   y  
2   republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   ?  
4   democrat   n   y   y   n   ?   y   n   n   n   n   y   n   y   n   n   y  
5   democrat   y   y   y   n   y   y   n   n   n   n   y   ?   y   y   y   y  
7   democrat   n   y   n   y   y   y   n   n   n   n   n   n   ?   y   y   y  
8   republican n   y   n   y   y   y   n   n   n   n   n   n   y   y   ?   y  
9   republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   y  
11  republican n   y   n   y   y   n   n   n   n   n   ?   ?   y   y   n   n  
12  republican n   y   n   y   y   y   n   n   n   n   y   ?   y   y   ?   ?  
13  democrat   n   y   y   n   n   n   y   y   y   n   n   n   y   n   ?   ?  
14  democrat   y   y   y   n   n   y   y   y   ?   y   y   ?   n   n   y   ?  
15  republican n   y   n   y   y   y   n   n   n   n   n   y   ?   ?   n   ?  
16  republican n   y   n   y   y   y   n   n   n   y   n   y   y   ?   n   ?  
17  democrat   y   n   y   n   n   y   n   y   ?   y   y   y   ?   n   n   y  
18  democrat   y   ?   y   n   n   n   y   y   y   n   n   n   y   n   y   y  
20  democrat   y   y   y   n   n   n   y   y   y   n   y   n   n   n   y   y  
21  democrat   y   y   y   n   n   ?   y   y   n   n   y   n   n   n   y   y  
22  democrat   y   y   y   n   n   n   y   y   y   n   n   n   ?   ?   y   y  
23  democrat   y   ?   y   n   n   n   y   y   y   n   n   ?   n   n   y   y  
24  democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   y   y  
25  democrat   y   n   y   n   n   n   y   y   y   n   n   n   n   n   y   ?  
26  democrat   y   n   y   n   n   n   y   y   y   y   n   n   n   n   y   y  
28  democrat   y   y   y   n   n   n   y   y   y   n   y   n   n   n   y   y  
29  republican y   n   n   y   y   n   y   y   y   n   n   y   y   y   n   y  
30  democrat   y   y   y   n   n   n   y   y   y   n   y   n   n   n   y   y  
31  republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   n  
32  democrat   y   y   y   n   n   n   y   y   y   n   y   n   n   n   y   ?  
33  democrat   y   y   y   n   n   n   y   y   y   y   n   n   y   n   y   y  
34  republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   y  
35  democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   y   y  
... ...        ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ...
390 democrat   y   n   y   n   n   n   y   y   y   ?   y   n   n   n   y   ?  
392 democrat   y   y   n   n   n   n   n   y   y   n   y   n   n   n   y   n  
394 republican ?   ?   ?   ?   n   y   n   y   y   n   n   y   y   n   n   ?  
395 democrat   y   y   ?   ?   ?   y   n   n   n   n   y   n   y   n   n   y  
396 democrat   y   y   y   ?   n   n   n   y   n   n   y   ?   n   n   y   y  
397 democrat   y   y   y   n   y   y   n   y   n   n   y   n   y   n   y   y  
398 democrat   y   y   n   n   y   ?   n   n   n   n   y   n   y   y   n   y  
399 democrat   n   y   y   n   y   y   n   y   n   n   n   n   n   n   n   y  
402 republican n   y   n   y   y   y   n   n   n   y   y   y   y   y   n   n  
404 republican n   y   n   y   y   y   n   n   n   y   n   y   ?   y   n   n  
407 democrat   y   n   y   n   y   y   n   n   y   y   n   n   y   y   n   y  
408 democrat   n   n   n   y   y   y   n   n   n   n   y   y   y   y   n   n  
410 republican n   n   n   y   y   y   n   n   n   n   n   y   y   y   n   n  
413 republican n   n   n   y   y   y   n   n   n   y   n   y   y   y   n   y  
415 democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   n   y  
416 democrat   n   y   y   n   n   y   y   y   ?   y   n   n   n   n   n   y  
417 republican y   y   n   y   y   y   n   n   n   y   n   n   y   y   n   y  
419 democrat   y   y   y   n   n   n   y   y   n   y   n   n   n   n   n   y  
420 democrat   y   y   y   n   n   n   y   y   y   n  

We use the rest of the model data to test our model;consequently, it is assigned to testdata 
variable. 

In [106]:
testdata<-X[-test,]
nrow(testdata)

[1] 305

Logistic regression model is formed by using glm function in R language. The meaning of glm is fitting  generalized linear models.

In [82]:
logitmodel<- glm(V1 ~ V4+V7+V9+V12+V15+V17, family=binomial(link="logit"), data= modeldata)
logitmodel

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Call:  glm(formula = V1 ~ V4 + V7 + V9 + V12 + V15 + V17, family = binomial(link = "logit"), 
    data = modeldata)

Coefficients:
(Intercept)          V4n          V4y          V7n          V7y          V9n  
     1.4788      20.6449      18.0192      -1.2764       0.1679     -36.9222  
        V9y         V12n         V12y         V15n         V15y         V17n  
   -37.7751      -0.4491      -2.5876      -1.8398      16.8952       0.2081  
       V17y  
     1.3671  

Degrees of Freedom: 129 Total (i.e. Null);  117 Residual
Null Deviance:	    174.1 
Residual Deviance: 58.02 	AIC: 84.02

After the model is generated, we use summary code to get a short information about the model 
that we generate.

In [83]:
summary(logitmodel)
anova(logitmodel, test="Chisq")


Call:
glm(formula = V1 ~ V4 + V7 + V9 + V12 + V15 + V17, family = binomial(link = "logit"), 
    data = modeldata)

Deviance Residuals: 
     Min        1Q    Median        3Q       Max  
-2.53898  -0.00007  -0.00003   0.39453   1.88076  

Coefficients:
              Estimate Std. Error z value Pr(>|z|)
(Intercept)     1.4788 22106.5346   0.000    1.000
V4n            20.6449  9546.5638   0.002    0.998
V4y            18.0192  9546.5638   0.002    0.998
V7n            -1.2764 17877.7056   0.000    1.000
V7y             0.1679 17877.7055   0.000    1.000
V9n           -36.9222 12828.0371  -0.003    0.998
V9y           -37.7751 12828.0371  -0.003    0.998
V12n           -0.4491     1.5932  -0.282    0.778
V12y           -2.5876     1.7423  -1.485    0.137
V15n           -1.8398  9584.1137   0.000    1.000
V15y           16.8952  9306.3907   0.002    0.999
V17n            0.2081     1.1858   0.175    0.861
V17y            1.3671     1.0004   1.367    0.172

(Dispersion parameter for bino

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"

Df Deviance  Resid. Df Resid. Dev Pr(>Chi)    
NULL NA        NA 129       174.13998            NA
V4    2 76.260184 127        97.87980  2.756205e-17
V7    2 11.134621 125        86.74518  3.820743e-03
V9    2  6.521919 123        80.22326  3.835157e-02
V12   2 10.529084 121        69.69417  5.171760e-03
V15   2  8.786661 119        60.90751  1.235950e-02
V17   2  2.885198 117        58.02231  2.363128e-01

After the model is formed, we use predict function to estimate the model for testdata variable 
which is the rest of the modeldata.

In [84]:
prelogitmodel<-predict(logitmodel,testdata, type = "response")

Now, `prelogitmodel` contains the probability that given MP is a republican or a democrat.  We need to set a threshold for the probability. If it is greater than 0.5 then we will predict that MP is a republican.

In [103]:
a<-ifelse(prelogitmodel>0.6,"republican","democrat")

In [104]:
table(a)

a
  democrat republican 
       199        106 

To compare logitmodel that we constitute and prelogitmodel that predict our model, we use
table function. We need to compare the number of democrat congressmen and republican congressmen.For this reason, we take V1 column of the testdata that consists of the class of the congressmen which are democrat and republican. Also, the table(a) that is formed by using prelogitmodel consists of the number of the congressmen that belongs to two class.

In [105]:
table(real=testdata$V1,predicted=a)

            predicted
real         democrat republican
  democrat        176         12
  republican       23         94

When we look at the anova function that gives informations about prelogitmodel, we take out laws
that their residual deviance has a small differences. Therefore, V9 and V17 are taken out from 
the code to change our model.
    

In [49]:
logitmodel<- glm(V1 ~ V4+V7+V12+V15, family=binomial(link="logit"), data= modeldata)
logitmodel

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Call:  glm(formula = V1 ~ V4 + V7 + V12 + V15, family = binomial(link = "logit"), 
    data = modeldata)

Coefficients:
(Intercept)          V4n          V4y          V7n          V7y         V12n  
   -18.5995      20.3249      17.5756      -0.1733       1.5369       0.1027  
       V12y         V15n         V15y  
    -2.2043     -19.6450      -0.7540  

Degrees of Freedom: 129 Total (i.e. Null);  121 Residual
Null Deviance:	    174.1 
Residual Deviance: 63.66 	AIC: 81.66

In [50]:
summary(logitmodel)
anova(logitmodel, test="Chisq")


Call:
glm(formula = V1 ~ V4 + V7 + V12 + V15, family = binomial(link = "logit"), 
    data = modeldata)

Deviance Residuals: 
     Min        1Q    Median        3Q       Max  
-2.31605  -0.00010  -0.00003   0.37650   2.24870  

Coefficients:
              Estimate Std. Error z value Pr(>|z|)
(Intercept)   -18.5995 20268.5386  -0.001    0.999
V4n            20.3249  9545.5644   0.002    0.998
V4y            17.5756  9545.5644   0.002    0.999
V7n            -0.1733 17880.0406   0.000    1.000
V7y             1.5369 17880.0406   0.000    1.000
V12n            0.1027     1.2782   0.080    0.936
V12y           -2.2043     1.3648  -1.615    0.106
V15n          -19.6450  2308.6386  -0.009    0.993
V15y           -0.7540     1.4698  -0.513    0.608

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 174.140  on 129  degrees of freedom
Residual deviance:  63.658  on 121  degrees of freedom
AIC: 81.658

Number of Fisher Scoring iterations: 19


Df Deviance Resid. Df Resid. Dev Pr(>Chi)    
NULL NA       NA 129       174.13998            NA
V4    2 76.26018 127        97.87980  2.756205e-17
V7    2 11.13462 125        86.74518  3.820743e-03
V12   2 12.29007 123        74.45511  2.144103e-03
V15   2 10.79707 121        63.65804  4.523201e-03

When we look at the table above, the differences between the residual deviances are close to each other. Consequently, we constitute our logitmodel again without V9 and V17 columns.

In [51]:
prelogitmodel<-predict(logitmodel,testdata, type = "response")

In [79]:
a<-ifelse(prelogitmodel>0.6,"republican","democrat")

In [80]:
table(a)

a
  democrat republican 
       215         90 

In [81]:
table(real=testdata$V1,predicted=a)

            predicted
real         democrat republican
  democrat        184          4
  republican       31         86